qiskit/passmanager/passmanager.py

```diff
@@ -12,183 +12,203 @@
 
 """Manager for a set of Passes and their scheduling during transpilation."""
 from __future__ import annotations
-from abc import ABC
+
+import logging
+import warnings
+from abc import ABC, abstractmethod
 from collections.abc import Callable, Sequence
 from typing import Any
-import logging
+
 import dill
-from qiskit.tools.parallel import parallel_map
 
-from .base_pass import GenericPass
-from .passrunner import BasePassRunner
+from qiskit.tools.parallel import parallel_map
+from .base_tasks import Task, BaseController, PassManagerIR
 from .exceptions import PassManagerError
-from .flow_controllers import FlowController, PassSequence
+from .flow_controllers import FlowControllerLinear, FlowController
+from .propertyset import PropertySet
 
 logger = logging.getLogger(__name__)
 
 
 class BasePassManager(ABC):
     """Pass manager base class."""
 
-    PASS_RUNNER = BasePassRunner
-
     def __init__(
         self,
-        passes: PassSequence | None = None,
+        tasks: list[Task] = None,
         max_iteration: int = 1000,
     ):
         """Initialize an empty pass manager object.
 
         Args:
-            passes: A pass set to be added to the pass manager schedule.
+            tasks: A pass set to be added to the pass manager schedule.
             max_iteration: The maximum number of iterations the schedule will be looped if the
                 condition is not met.
         """
-        # the pass manager's schedule of passes, including any control-flow.
-        # Populated via PassManager.append().
-        self._pass_sets: list[dict[str, Any]] = []
+        self._flow_controller = FlowControllerLinear()
         self.max_iteration = max_iteration
+        self.property_set = PropertySet()
 
-        if passes is not None:
-            self.append(passes)
+        # For backward compatibility.
+        # This wraps pass list with a linear flow controller.
+        if tasks is not None:
+            self.append(tasks)
```

Comment:
I don't think this architecture of storing a `_flow_controller` works quite right logically.  `BasePassManager.append` mutates an inner attribute of `FlowControllerLinear` that's made a `tuple` here, which makes building a `PassManager` with `append` have quadratic complexity and breaks the API design that a `FlowController` is meant to be constructed with all its passes at once.

I think it might be better to _not_ have `_flow_controller` as state on the pass manager, and instead have a `self._tasks: List[Task]`, which during `BasePassManager.run` we wrap into a `FlowControllerLinear` if that's the way we want to go.  That lets us have fast appends, maintain the flow-controller API safely, and still re-use the execution logic from `FlowControllerLinear` to do the run.

Reply:
You're absolutely correct. With this change we can very reasonably remove `.append` from the controllers. Done in [b9348ae](https://github.com/Qiskit/qiskit/pull/10127/commits/b9348ae5ed6364a860b069978686fd5fe95dd70c)